src/main/java/appeng/block/AEBaseTileBlock.java

```diff
@@ -271,87 +271,93 @@ public void onBlockPlacedBy( final World w, final BlockPos pos, final IBlockStat
-						{
-							return false;
-						}
+				if( tile == null )
+				{
+					return false;
+				}
 
-						final ItemStack op = new ItemStack( this );
-						for( final ItemStack ol : drops )
-						{
-							if( Platform.isSameItemType( ol, op ) )
-							{
-								final NBTTagCompound tag = tile.downloadSettings( SettingsFrom.DISMANTLE_ITEM );
-								if( tag != null )
-								{
-									ol.setTagCompound( tag );
-								}
-							}
-						}
+				if( tile instanceof TileCableBus || tile instanceof TileSkyChest )
+				{
+					return false;
+				}
+
+				final ItemStack[] itemDropCandidates = Platform.getBlockDrops( w, pos );
+				final ItemStack op = new ItemStack( this );
 
-						if( id.removedByPlayer( ids, w, pos, player, false ) )
+				for( final ItemStack ol : itemDropCandidates )
+				{
+					if( Platform.isSameItemType( ol, op ) )
+					{
+						final NBTTagCompound tag = tile.downloadSettings( SettingsFrom.DISMANTLE_ITEM );
+						if( tag != null )
 						{
-							final List<ItemStack> l = Lists.newArrayList( drops );
-							Platform.spawnDrops( w, pos, l );
-							w.setBlockToAir( pos );
+							ol.setTagCompound( tag );
 						}
 					}
+				}
+
+				if( block.removedByPlayer( blockState, w, pos, player, false ) )
+				{
+					final List<ItemStack> itemsToDrop = Lists.newArrayList( itemDropCandidates );
+					Platform.spawnDrops( w, pos, itemsToDrop );
+					w.setBlockToAir( pos );
+				}
+
+				return false;
+			}
+
+			if( heldItem.getItem() instanceof IMemoryCard && !( this instanceof BlockCableBus ) )
+			{
+				final IMemoryCard memoryCard = (IMemoryCard) heldItem.getItem();
+				final AEBaseTile tileEntity = this.getTileEntity( w, pos );
+
+				if( tileEntity == null )
+				{
 					return false;
 				}
 
-				if( heldItem.getItem() instanceof IMemoryCard && !( this instanceof BlockCableBus ) )
+				final String name = this.getUnlocalizedName();
+
+				if( player.isSneaking() )
 				{
-					final IMemoryCard memoryCard = (IMemoryCard) heldItem.getItem();
-					if( player.isSneaking() )
+					final NBTTagCompound data = tileEntity.downloadSettings( SettingsFrom.MEMORY_CARD );
+					if( data != null )
 					{
-						final AEBaseTile t = this.getTileEntity( w, pos );
-						if( t != null )
-						{
-							final String name = this.getUnlocalizedName();
-							final NBTTagCompound data = t.downloadSettings( SettingsFrom.MEMORY_CARD );
-							if( data != null )
-							{
-								memoryCard.setMemoryCardContents( heldItem, name, data );
-								memoryCard.notifyUser( player, MemoryCardMessages.SETTINGS_SAVED );
-								return true;
-							}
-						}
+						memoryCard.setMemoryCardContents( heldItem, name, data );
+						memoryCard.notifyUser( player, MemoryCardMessages.SETTINGS_SAVED );
+					}
+				}
+				else
+				{
+					final String savedName = memoryCard.getSettingsName( heldItem );
+					final NBTTagCompound data = memoryCard.getData( heldItem );
+
+					if( this.getUnlocalizedName().equals( savedName ) )
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Refactor this code to not nest more than 3 if/for/while/switch/try statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AS134)